File: src/lib/matching/group.ts
// Optimized group matching: weighted scoring, distance decay, pre-computed Jaccard.
// Industry patterns: content-based filtering, distance decay, weighted multi-factor scoring.

// --- 1. DATA TYPE DEFINITIONS ---

interface Location {
  lat: number;
  lon: number;
}

export interface UserProfile {
  userId: string;
  destination: Location;
  budget: number;
  startDate: string; // ISO: YYYY-MM-DD
  endDate: string; // ISO: YYYY-MM-DD
  age: number;
  languages: string[];
  interests: string[];
  smoking: boolean;
  drinking: boolean;
  nationality: string;
}

export interface GroupProfile {
  groupId: string;
  name: string;
  destination: Location;
  averageBudget: number;
  startDate: string;
  endDate: string;
  averageAge: number;
  dominantLanguages: string[];
  topInterests: string[];
  smokingPolicy: "Smokers Welcome" | "Mixed" | "Non-Smoking";
  drinkingPolicy: "Drinkers Welcome" | "Mixed" | "Non-Drinking";
  dominantNationalities: string[];
  /** Optional: pre-computed distance in km for distance decay scoring */
  distanceKm?: number;
}

// --- 2. HELPER & SCORING FUNCTIONS ---

/** Haversine distance in km (optimized, single computation) */
const haversineKm = (a: Location, b: Location): number => {
  const R = 6371;
  const dLat = (b.lat - a.lat) * (Math.PI / 180);
  const dLon = (b.lon - a.lon) * (Math.PI / 180);
  const x =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(a.lat * (Math.PI / 180)) *
      Math.cos(b.lat * (Math.PI / 180)) *
      Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(x), Math.sqrt(1 - x));
};

const isWithinDistance = (
  loc1: Location,
  loc2: Location,
  maxDistanceKm: number = 200,
): boolean => haversineKm(loc1, loc2) <= maxDistanceKm;

/**
 * SCORE: Calculates budget similarity from 0 to 1.
 */
const calculateBudgetScore = (
  userBudget: number,
  groupAverageBudget: number,
): number => {
  if (groupAverageBudget <= userBudget) return 1.0;
  const budgetDifference = groupAverageBudget - userBudget;
  // Score reaches 0 if the difference is 40k or more.
  return Math.max(0, 1 - budgetDifference / 40000);
};

/**
 * SCORE: Calculates date overlap score from 0 to 1.
 */
const calculateDateOverlapScore = (
  userStart: string,
  userEnd: string,
  groupStart: string,
  groupEnd: string,
): number => {
  const uStart = new Date(userStart).getTime();
  const uEnd = new Date(userEnd).getTime();
  const gStart = new Date(groupStart).getTime();
  const gEnd = new Date(groupEnd).getTime();

  const overlapStart = Math.max(uStart, gStart);
  const overlapEnd = Math.min(uEnd, gEnd);

  if (overlapStart > overlapEnd) return 0;

  const overlapDuration = (overlapEnd - overlapStart) / (1000 * 60 * 60 * 24) + 1;
  const userTripDuration = (uEnd - uStart) / (1000 * 60 * 60 * 24) + 1;

  return Math.min(1, overlapDuration / userTripDuration);
};

/**
 * SCORE: Calculates age similarity score.
 */
const calculateAgeScore = (userAge: number, groupAverageAge: number): number => {
  if (!groupAverageAge) return 0.5;
  const ageDifference = Math.abs(userAge - groupAverageAge);
  return Math.max(0, 1 - ageDifference / 20);
};

/** Jaccard similarity using pre-computed Sets. */
const jaccardFromSets = (setA: Set<string>, setB: Set<string>): number => {
  if (setA.size === 0 && setB.size === 0) return 0.5;
  if (setA.size === 0 || setB.size === 0) return 0.5;
  let intersection = 0;
  const smaller = setA.size <= setB.size ? setA : setB;
  const larger = setA.size <= setB.size ? setB : setA;
  for (const x of smaller) if (larger.has(x)) intersection++;
  const union = setA.size + setB.size - intersection;
  return union === 0 ? 0.5 : intersection / union;
};

/**
 * SCORE: Calculates lifestyle compatibility (smoking & drinking).
 */
const calculateLifestyleScore = (
  user: UserProfile,
  group: GroupProfile,
): number => {
  let smokingScore = 0;
  if (
    (user.smoking && group.smokingPolicy === "Smokers Welcome") ||
    (!user.smoking && group.smokingPolicy === "Non-Smoking")
  ) {
    smokingScore = 1.0;
  } else if (group.smokingPolicy === "Mixed") {
    smokingScore = 0.6;
  }

  let drinkingScore = 0;
  if (
    (user.drinking && group.drinkingPolicy === "Drinkers Welcome") ||
    (!user.drinking && group.drinkingPolicy === "Non-Drinking")
  ) {
    drinkingScore = 1.0;
  } else if (group.drinkingPolicy === "Mixed") {
    drinkingScore = 0.6;
  }

  return (smokingScore + drinkingScore) / 2;
};

const calculateBackgroundScore = (
  user: UserProfile,
  group: GroupProfile,
): number => {
  if (
    !user.nationality ||
    user.nationality === "Unknown" ||
    user.nationality === "Any"
  )
    return 0.5;
  if (!group.dominantNationalities?.length) return 0.5;
  return group.dominantNationalities.some(
    (n) => n.toLowerCase() === user.nationality.toLowerCase(),
  )
    ? 1.0
    : 0.0;
};

/** Distance decay boost */
const distanceDecayScore = (distanceKm: number, maxKm: number): number => {
  if (distanceKm <= 0) return 1;
  if (distanceKm >= maxKm) return 0;
  return 1 - distanceKm / maxKm;
};

// --- 3. CORE MATCHING FUNCTION ---

/**
 * Finds the best group matches using weighted multi-factor scoring and ML.
 */
export const findGroupMatchesForUser = async (
  user: UserProfile,
  allGroups: GroupProfile[],
  maxDistanceKm: number = 200,
  /** @deprecated distance filtering should be done before calling this function */
  _deprecatedMaxDistanceKm?: number,
  useML: boolean = true,
) => {
  // Pre-compute user sets for efficiency
  const userLangSet = new Set(
    (user.languages || []).map((s) => s.toLowerCase().trim()).filter(Boolean),
  );
  const userInterestSet = new Set(
    (user.interests || []).map((s) => s.toLowerCase().trim()).filter(Boolean),
  );

  const weights = {
    budget: 0.18,
    dateOverlap: 0.2,
    interests: 0.15,
    age: 0.12,
    language: 0.1,
    lifestyle: 0.1,
    background: 0.05,
    distance: 0.1,
  };

  let mlScores = new Map<string, number | null>();
  if (useML && allGroups.length > 0) {
    try {
      const { calculateMLGroupCompatibilityScoresBatch } = await import("../ai/matching/ml-scoring");
      mlScores = await calculateMLGroupCompatibilityScoresBatch(
        {
          destination: user.destination,
          startDate: user.startDate,
          endDate: user.endDate,
          budget: user.budget,
          age: user.age,
          interests: user.interests,
          languages: user.languages,
          smoking: user.smoking ? "yes" : "no",
          drinking: user.drinking ? "yes" : "no",
          nationality: user.nationality,
        },
        allGroups,
        { enabled: true, fallbackOnError: true }
      );
    } catch (error) {
      console.debug("ML group batch scoring unavailable:", error);
    }
  }

  const scoredMatches = await Promise.all(
    allGroups.map(async (group) => {
      const budgetScore = calculateBudgetScore(user.budget, group.averageBudget);
      const dateOverlapScore = calculateDateOverlapScore(
        user.startDate,
        user.endDate,
        group.startDate,
        group.endDate,
      );

      const groupInterestSet = new Set(
        (group.topInterests || []).map((s) => String(s).toLowerCase().trim()),
      );
      const groupLangSet = new Set(
        (group.dominantLanguages || []).map((s) =>
          String(s).toLowerCase().trim(),
        ),
      );

      const interestScore = jaccardFromSets(userInterestSet, groupInterestSet);
      const ageScore = calculateAgeScore(user.age, group.averageAge);
      const languageScore = jaccardFromSets(userLangSet, groupLangSet);
      const lifestyleScore = calculateLifestyleScore(user, group);
      const backgroundScore = calculateBackgroundScore(user, group);
      const distanceScore =
        group.distanceKm !== undefined
          ? distanceDecayScore(group.distanceKm, maxDistanceKm)
          : 1;

      const ruleBasedScore =
        budgetScore * weights.budget +
        dateOverlapScore * weights.dateOverlap +
        interestScore * weights.interests +
        ageScore * weights.age +
        languageScore * weights.language +
        lifestyleScore * weights.lifestyle +
        backgroundScore * weights.background +
        distanceScore * weights.distance;

      let finalScore = ruleBasedScore;
      const mlScore = mlScores.get(group.groupId) ?? null;

      if (typeof mlScore === "number" && !isNaN(mlScore)) {
        // Blend ML score (70%) with rule-based score (30%)
        finalScore = mlScore * 0.7 + ruleBasedScore * 0.3;
      } else {
        // Fallback to 100% rule-based if ML failed
        finalScore = ruleBasedScore;
      }

      // Final safety guard
      if (isNaN(finalScore) || !isFinite(finalScore)) {
        finalScore = ruleBasedScore || 0.5;
      }

      return {
        group,
        score: parseFloat(finalScore.toFixed(3)),
        mlScore: mlScore !== null ? parseFloat(mlScore.toFixed(3)) : undefined,
        breakdown: {
          budget: parseFloat(budgetScore.toFixed(3)),
          dates: parseFloat(dateOverlapScore.toFixed(3)),
          interests: parseFloat(interestScore.toFixed(3)),
          age: parseFloat(ageScore.toFixed(3)),
        },
      };
    }),
  );

  return scoredMatches.sort((a, b) => b.score - a.score);
};

File: apps/matching-service/internal/matching/group.go
package matching

import (
	"math"
	"strings"

	"github.com/kovari/matching-service/internal/models"
)

func CalculateGroupBudgetScore(userBudget, groupAvgBudget float64) float64 {
	if groupAvgBudget <= userBudget {
		return 1.0
	}
	diff := groupAvgBudget - userBudget
	return math.Max(0, 1-diff/40000.0)
}

func CalculateGroupAgeScore(userAge int, groupAvgAge float64) float64 {
	if groupAvgAge == 0 {
		return 0.5
	}
	diff := math.Abs(float64(userAge) - groupAvgAge)
	return math.Max(0, 1-diff/20.0)
}

func CalculateGroupLifestyleScore(user models.SoloSession, group models.GroupProfile) float64 {
	uA := user.StaticAttributes
	smokingScore := 0.0
	uSmoke := uA != nil && strings.EqualFold(uA.Smoking, "yes")
	if (uSmoke && group.SmokingPolicy == "Smokers Welcome") || (!uSmoke && group.SmokingPolicy == "Non-Smoking") {
		smokingScore = 1.0
	} else if group.SmokingPolicy == "Mixed" {
		smokingScore = 0.6
	}

	drinkingScore := 0.0
	uDrink := uA != nil && strings.EqualFold(uA.Drinking, "yes")
	if (uDrink && group.DrinkingPolicy == "Drinkers Welcome") || (!uDrink && group.DrinkingPolicy == "Non-Drinking") {
		drinkingScore = 1.0
	} else if group.DrinkingPolicy == "Mixed" {
		drinkingScore = 0.6
	}

	return (smokingScore + drinkingScore) / 2.0
}

func CalculateGroupBackgroundScore(user models.SoloSession, group models.GroupProfile) float64 {
	uA := user.StaticAttributes
	if uA == nil || uA.Nationality == "" || uA.Nationality == "Unknown" || uA.Nationality == "Any" {
		return 0.5
	}
	for _, n := range group.DominantNationalities {
		if strings.EqualFold(n, uA.Nationality) {
			return 1.0
		}
	}
	return 0.0
}

func CalculateDistanceDecayScore(distanceKm, maxKm float64) float64 {
	if distanceKm <= 0 {
		return 0.5 // Neutral/Unknown fallback
	}
	if distanceKm >= maxKm {
		return 0
	}
	return 1.0 - (distanceKm / maxKm)
}

func CalculateGroupDateOverlapScore(start1, end1, start2, end2 string) float64 {
	overlapDays, tripDuration := CalculateDateOverlapGroup(start1, end1, start2, end2)
	if overlapDays < 1 || tripDuration <= 0 {
		return 0
	}
	return math.Min(1.0, overlapDays/tripDuration)
}

func CalculateFinalGroupScore(user models.SoloSession, group models.GroupProfile, mlScore *float64, config *models.MatchingConfig) models.GroupMatchResult {
	budgetScore := CalculateGroupBudgetScore(user.Budget, group.AverageBudget)
	dateOverlapScore := CalculateGroupDateOverlapScore(user.StartDate, user.EndDate, group.StartDate, group.EndDate)
	
	var interestScore, ageScore, languageScore float64
	uA := user.StaticAttributes
	if uA != nil {
		interestScore = CalculateJaccardSimilarity(uA.Interests, group.TopInterests)
		ageScore = CalculateGroupAgeScore(uA.Age, group.AverageAge)
		languageScore = CalculateJaccardSimilarity(uA.Languages, group.DominantLanguages)
	} else {
		// Minimum penalty if user profile missing at this stage (should have been filtered)
		interestScore, ageScore, languageScore = 0.3, 0.5, 0.3
	}
	
	lifestyleScore := CalculateGroupLifestyleScore(user, group)
	backgroundScore := CalculateGroupBackgroundScore(user, group)
	distanceScore := CalculateDistanceDecayScore(group.DistanceKm, 1500.0)

	weights := config.GroupWeights
	ruleBasedScore := budgetScore*weights["budget"] +
		dateOverlapScore*weights["dates"] +
		interestScore*weights["interests"] +
		ageScore*weights["age"] +
		languageScore*weights["language"] +
		lifestyleScore*weights["lifestyle"] +
		backgroundScore*weights["background"] +
		distanceScore*weights["distance"]

	finalScore := ruleBasedScore
	if mlScore != nil {
		adjustedMl := *mlScore
		// Non-linear boost for conservative group ML scores
		if adjustedMl > 0 && adjustedMl < 0.3 {
			adjustedMl *= 3.5
		} else if adjustedMl >= 0.3 && adjustedMl < 0.6 {
			adjustedMl *= 1.5
		}
		if adjustedMl > 0.95 {
			adjustedMl = 0.95
		}

		mlBlend := config.MLBlend["group"]
		finalScore = adjustedMl*mlBlend + ruleBasedScore*(1.0-mlBlend)
	}

	// Sum then round to match TS exactly
	finalScore = math.Round(finalScore*1000) / 1000

	return models.GroupMatchResult{
		Group: group,
		Score: finalScore,
		MLScore: mlScore,
		Breakdown: map[string]float64{
			"budget":    math.Round(budgetScore*1000) / 1000,
			"dates":     math.Round(dateOverlapScore*1000) / 1000,
			"interests": math.Round(interestScore*1000) / 1000,
			"age":       math.Round(ageScore*1000) / 1000,
		},
	}
}
